push_notifications/apns_async.py

```diff
@@ -0,0 +1,343 @@
+	"""
+	The key for a localized subtitle string. Use this key, instead of the subtitle key, to retrieve the subtitle from your app’s Localizable.strings file. The value must contain the name of a key in your strings file.
+	"""
+
+	subtitle_loc_args: list[str] = NotSet
+	"""
+	An array of strings containing replacement values for variables in your title string. Each %@ character in the string specified by subtitle-loc-key is replaced by a value from this array. The first item in the array replaces the first instance of the %@ character in the string, the second item replaces the second instance, and so on.
+	"""
+
+	loc_key: str = NotSet
+	"""
+	The key for a localized message string. Use this key, instead of the body key, to retrieve the message text from your app’s Localizable.strings file. The value must contain the name of a key in your strings file.
+	"""
+
+	loc_args: list[str] = NotSet
+	"""
+	An array of strings containing replacement values for variables in your message text. Each %@ character in the string specified by loc-key is replaced by a value from this array. The first item in the array replaces the first instance of the %@ character in the string, the second item replaces the second instance, and so on.
+	"""
+
+	sound: Union[str, any] = NotSet
+	"""
+	string
+	The name of a sound file in your app’s main bundle or in the Library/Sounds folder of your app’s container directory. Specify the string “default” to play the system sound. Use this key for regular notifications. For critical alerts, use the sound dictionary instead. For information about how to prepare sounds, see UNNotificationSound.
+
+	dictionary
+	A dictionary that contains sound information for critical alerts. For regular notifications, use the sound string instead.
+	"""
+
+	def asDict(self) -> dict[str, any]:
+		python_dict = asdict(self)
+		return {
+			key.replace("_", "-"): value
+			for key, value in python_dict.items()
+			if value is not NotSet
+		}
+
+
+class APNsService:
+	__slots__ = ("client",)
+
+	def __init__(
+		self, application_id: str = None, creds: Credentials = None, topic: str = None
+	):
+		try:
+			loop = asyncio.get_event_loop()
+		except RuntimeError:
+			loop = asyncio.new_event_loop()
+			asyncio.set_event_loop(loop)
+
+		self.client = self._create_client(
+			creds=creds, application_id=application_id, topic=topic
+		)
+
+	def send_message(
+		self,
+		request: NotificationRequest,
+	):
+		loop = asyncio.get_event_loop()
+		res1 = self.client.send_notification(request)
+		res = loop.run_until_complete(res1)
+		return res
+
+	def _create_notification_request_from_args(
+		self,
+		registration_id: str,
+		alert: Union[str, Alert],
+		badge: int = None,
+		sound: str = None,
+		extra: dict = {},
+		expiration: int = None,
+		thread_id: str = None,
+		loc_key: str = None,
+		priority: int = None,
+		collapse_id: str = None,
+		aps_kwargs: dict = {},
+		message_kwargs: dict = {},
+		notification_request_kwargs: dict = {},
```

Comment:
Since `notification_request_kwargs` argument has a default value and it's content is changed in the function logic, the default state can change during execution. If any previous call has set values `expiration`, `priority` or `collapse_id`, these are persisted to any next call and if the next call  does not have these values or has the value set to `None`, the values are still passed to sent `NotificationRequest`. A simple test case to demonstrate this is to send first message with `expiration` and second without `expiration` set. The value from first call is passed to second call `NotificationRequest`.

Reply:
Solved it by creating copy of `notification_request_kwargs`, but if you consider it code smell I could rework it to having default argument `None` and replacing them in function with `{}`. Added your test (`test_apns_send_messages_different_priority`)